app/controllers/upvotes_controller.rb

```diff
@@ -0,0 +1,17 @@
+class UpvotesController < ApplicationController
+  def create
+    @review = Review.find(params[:review_id])
+    @upvote = @review.upvotes.build(user: current_user)
+    if current_user.has_downvoted?(@review) && @upvote.save
+      downvote = current_user.downvotes.find_by(review: @review)
+      downvote.destroy
+      flash[:notice] = "New vote recorded"
+      redirect_to review_path(@review)
+    elsif @upvote.save
+      flash[:notice] = "Vote recorded"
+      redirect_to review_path(@review)
+    else
+      render "reviews#show"
+    end
+  end
+end
```

Comment:
There's a lot of duplicate code in the controller and model between upvotes and downvotes. Why not combine them into the same model? Upvotes and Downvotes are special cases of a `Vote` model. You can add an additional field to the table to indicate whether it is an upvote or downvote. It also simplifies the case of checking if a user has already voted on a review since the rows are in the same table.